lib/spack/spack/fetch_strategy.py

```diff
@@ -1575,11 +1575,11 @@ def for_package_version(pkg, version):
 
     check_pkg_attributes(pkg)
 
-    if not isinstance(version, spack.version.Version):
+    if not isinstance(version, spack.version.VersionBase):
         version = spack.version.Version(version)
 
     # if it's a commit, we must use a GitFetchStrategy
-    if version.is_commit and hasattr(pkg, "git"):
+    if isinstance(version, spack.version.GitVersion) and hasattr(pkg, "git"):
```

Comment:
Also I think this will fail for `git.master` versions. The `kwargs` for the fetch strategy are hard coded to use `commit` below this.

Should add tests for the different `git.[ref types]` that are supported.

Reply:
Modified it to raise if there is no 'git' attribute.

Treating branches/tags as commits wasn't going to cause a problem, it just has more conservative fetch semantics. I've changed it to separate by ref type.